src/raft/raft.go

```diff
@@ -151,13 +379,199 @@ type RequestVoteArgs struct {
+
+	DPrintf(LOG_LEVEL_ELECTION, "Host %d (%s, term: %d) recv RequestVoteRPC from (pid: %d, term: %d) ", rf.me, GetServerState(rf.nodeStatus), rf.currentTerm, args.CandidateId, candidate_term)
+ 
+	if candidate_term > rf.currentTerm {
+		DebugP(dElection,"S%d (%s) term is lower than candidate, reset to follower", rf.me, GetServerState(rf.nodeStatus))
+		rf.setStateToFollower(candidate_term)
+	}
+
+	reply.Term = host_term
+
+	if candidate_term < host_term {
+		reply.VoteGranted = false 
+	} else {
+ 
+		if rf.nodeStatus == Follower && (rf.votedFor == HAS_NOT_VOTED || rf.votedFor == args.CandidateId) && rf.candidateLogIsUpToDate(args.LastLogIndex, args.LastLogTerm, candidate_term) {
+			 
+			rf.votedFor = args.CandidateId
+			reply.VoteGranted = true
+			rf.lastContactWithPeer = time.Now()
+			rf.persist()
+			DebugP(dElection,"S%d granted vote for S%d, term %d", rf.me, rf.votedFor, rf.currentTerm)
+		}
+	}
+
+	rf.mu.Unlock()
+}
+
+
+func (rf * Raft) candidateLogIsUpToDate(candidateLastLogIndex int, candidateLastLogTerm int, candidateTerm int) (bool) {
+	hostLastLogTerm := rf.getLastLogTerm()
+	hostLastLogIdx := rf.getLastLogIdx()
+
+	if candidateLastLogTerm < hostLastLogTerm {
+		return false
+	}
+
+	if candidateLastLogTerm > hostLastLogTerm {
+		return true
+	}
+
+	if candidateLastLogTerm == hostLastLogTerm {
+		if candidateLastLogIndex < hostLastLogIdx {
+			return false
+		}
+	}
+
+	return true;
+}
+
+func (rf *Raft) AppendEntries(args *AppendEntriesArgs, reply *AppendEntriesReply) {
+	
+	rf.mu.Lock()
+  	candidate_term := args.Term
+	host_term := rf.currentTerm
+ 
+ 
+	// check_result := rf.termCheck(candidate_term)
+	reply.Term  = rf.currentTerm
+	reply.LogConsistent = true
+	reply.Xterm = -1
+	reply.XIndex = -1
+	reply.Xlen = -1
+
+	if candidate_term < host_term {
+		DebugP(dElection,"Reject AE from Leader %d due to term mismatch", args.LeaderId)
+		reply.Success = false
+		rf.mu.Unlock()
+		return 
+	}
+
+	if candidate_term > host_term {
+		rf.setStateToFollower(candidate_term)
+		rf.leaderId = args.LeaderId
+	}
+
+	 if rf.serverContainsLeaderLog(args.PrevLogIndex, args.PrevLogTerm) == false {
+		rf.lastContactWithPeer = time.Now()
+		rf.leaderId = args.LeaderId
+		reply.Success = false
+		reply.LogConsistent = false
+ 
+		if rf.getLastLogIdx() < args.PrevLogIndex {
+			reply.Xlen = len(rf.logs) + rf.lastIncludedIdx
+		} else {
+			conflict_term := rf.getLogTermAtIndex(args.PrevLogIndex)
+			first_idx_with_conflict := rf.lookupFirstEntryWithTerm(conflict_term)
+
+			reply.Xterm = conflict_term
+			reply.XIndex = first_idx_with_conflict
+		}
+
+	} else if rf.serverContainsLeaderLog(args.PrevLogIndex, args.PrevLogTerm) == true {
+ 
+		rf.lastContactWithPeer = time.Now()
+		rf.leaderId = args.LeaderId
+		
+		reply.LogConsistent = true
+		reply.Success = true
+
+		rf.acceptLogsFromLeader(&args.Entries, args.PrevLogIndex + 1)
+ 
+		if args.LeaderCommit > rf.commitIndex {
+			lastNewLogIdx := args.PrevLogIndex + len(args.Entries)
```

Comment:
This was a rather perplexing bug, and it ultimately came down to the mis-interpretation of Figure 2, step 5  of the `AppendEntries` RPC implementation. 

The figure instructs the follower to update its commit index by set it to the min between `args.LeaderCommit` and `index of last new entry`. On the first pass the implementer may just use `rf.getLastLogIdx()` in all situations to get the `index of last new entry` but this isn't safe.

Consider the following situation, where we have a cluster of 5 servers S0 -> S4. S0 is the leader with logs [1,2,3,4,5,6], commitIndex=3, and S3 is the follower with logs [1,2,3,4,5,24,1204,305], commitIndex=2. Its logs [24,1204,305] were from leaders from another term that either crashed or was usurped.  S0 notices that 2 other servers, S2, and S4 have logs agreeing up to some index 6, and `S0.logs[6].CommandTerm == S0.currentTerm`. So it sets its commitIndex to 6.

S0 then sends out heart beat `AppendEntries `RPC where `args.Entries` is empty to S3. S3 do all the checks listed in figure 2. In previous iterations, S3 checks min(leaderCommit, rf.getLastLogIdx()). leaderCommit index is 6, and getLastLogIdx() will return 8. Thus, S3 updates it commit index, and erroneously apply its log entries at index 6, and the consensus algo breaks. 

To address this bug, we need to rethink the semantics of what `index of last new entry means`. In this case, the follower and leader agree up to, and including some index `args.PrevLogIndex`. `rf.acceptLogsFromLeader` will attempt to converge the follower's log with the leader's log based on the entries found in `args. Entries`.  

Then it must be that after calling `rf.acceptLogsFromLeader` , the follower must contain the same log entries as the leader up to some index `args.PrevLogIdx + len(args.Entries)`.

So in this example, S3 will update its commit index by running min(args.LeaderIndex, args.PrevLogIndex + len(rf.logs)) = min(6, 5+len([])) = 5.